application/models/services/UserManager.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace LimeSurvey\Models\Services;
+
+use LSWebUser;
+use Permission;
+use User;
+
+/**
+ * Service class for managing users and their permissions
+ */
+class UserManager
+{
+    /** @var LSWebUser the user managing other users */
+    private $managingUser;
+
+    /** @var User the user being handled */
+    private $targetUser;
+
+    /**
+     * @param LSWebUser $managingUser
+     * @param User|null $targetUser
+     */
+    public function __construct(
+        LSWebUser $managingUser,
+        User $targetUser = null
+    ) {
+        $this->managingUser = $managingUser;
+        $this->targetUser = $targetUser;
+    }
+
+    /**
+     * Returns true if the managing user can assign permissions to the target user.
+     * @return boolean
+     */
+    public function canAssignPermissions()
```

Comment:
? Not related here …

Reply:
Well, it is wrapping some common behaviours... That's the relationship...

Reply:
I wonder if this function wouldn't be clearer if the two users were given as argument instead of dependencies in the constructor. "Returns true if $user1 can assign permissions to $user2".

Reply:
Well.... That would turn things into a helper, and we are trying to move away from helpers.
It was out first choice, but trashed it after thinking you were not gonna like it :)

Also, one of the users is usually the same (the one assigning permissions or performing the actions), so why sending it always as a parameter.

Reply:
> I wonder if this function wouldn't be clearer if the two users were given as argument instead of dependencies in the constructor. "Returns true if $user1 can assign permissions to $user2".

Somethink like `hasPermssion` function …

Reply:
I understand `hasPermission` doesn't check if the user applying the permission has access to the target user.
`$this->targetUser->parent_id == $this->managingUser->id`

Reply:
> That would turn things into a helper, and we are trying to move away from helpers. It was out first choice, but trashed it after thinking you were not gonna like it :)
> 
> Also, one of the users is usually the same (the one assigning permissions or performing the actions), so why sending it always as a parameter.

Well. You shouldn't fit a round block into a square hole. :) Helper functions aren't wrong, when they _are_ helper functions, that is, without dependencies, (preferably) without side-effects.

It's easier to read. The function depicts a relationship between two entities. But where are the two entities? Somewhere else in the code, that I now have to find...